src/server/engine_shard_set.cc

```diff
@@ -56,6 +318,31 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+void EngineShard::StartDefragTask(util::ProactorBase* pb) {
+  MemoryDefragParams params(mi_resource_.heap(), &db_slice_);
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  defrag_task_ = pb->AddOnIdleTask(
```

Comment:
the main task - check if defrag is required, and if so, run it. return 0 if no defrag require, which will let this task "rest" for some time.